### PR TITLE
GPU Matrix

### PR DESCRIPTION
**Pull request by SteveBronder**
#### Submission Checklist

- [ ] Run unit tests: `./runTests.py test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary:

Adds the functions to move matrices to and from the GPU. The two main files are

1. `basic_matrix_kernels.hpp`
    - The GPU kernel for copying matrices on the GPU

2. `matrix_gpu.hpp`
    - `matrix_gpu` class for copying and creating matrices on the GPU

#### Intended Effect:

Allows for matrices to be passed to and from the GPU.

#### How to Verify:

```bash
echo STAN_OPENCL=true >> make/local
echo OPENCL_PLATFORM_ID=0 >> make/local
echo OPENCL_DEVICE_ID=0 >> make/local
make test/unit/math/gpu/opencl_matrix_test && ./test/unit/math/gpu/opencl_matrix_test
```

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

Rok Češnovar and Erik Štrumbelj (Faculty of Computer and Information Science, University of Ljubljana)
Steve Bronder

### CODE REVIEW

File: stan/math/gpu/opencl_context.hpp [hunk 1]
```diff
@@ -101,16 +103,32 @@ class opencl_context_base {
...
+    std::ifstream kernel_buff(file_name);
+    static std::string kernel;
+    kernel_buff.seekg(0, std::ios::end);
+    kernel.reserve(kernel_buff.tellg());
+    kernel_buff.seekg(0, std::ios::beg);
+    kernel.assign((std::istreambuf_iterator<char>(kernel_buff)),
+                  std::istreambuf_iterator<char>());
+    return kernel.c_str();
+  }
+
```
Comment (SteveBronder) [inline stan/math/gpu/opencl_context.hpp]: The static keyword at line 110 here is because I was having an initialization order fiasco on the kernel string without it.

**Comment by syclik** [comment]
Bad news: I ran it locally. It failed.
```
$ ./test/unit/math/gpu/opencl_matrix_test
Running main() from gtest_main.cc
[==========] Running 3 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 3 tests from MathMatrixGPU
[ RUN      ] MathMatrixGPU.matrix_gpu_creation
[       OK ] MathMatrixGPU.matrix_gpu_creation (17 ms)
[ RUN      ] MathMatrixGPU.matrix_gpu_copy
test/unit/math/gpu/opencl_matrix_test.cpp:48: Failure
Value of: d1_b(1)
  Actual: 0
Expected: 2
test/unit/math/gpu/opencl_matrix_test.cpp:49: Failure
Value of: d1_b(2)
  Actual: 0
Expected: 3
test/unit/math/gpu/opencl_matrix_test.cpp:65: Failure
Value of: d2_b(0, 2)
  Actual: 0
Expected: 3
test/unit/math/gpu/opencl_matrix_test.cpp:67: Failure
Value of: d2_b(1, 1)
  Actual: 0
Expected: 5
test/unit/math/gpu/opencl_matrix_test.cpp:68: Failure
Value of: d2_b(1, 2)
  Actual: 0
Expected: 6
[  FAILED  ] MathMatrixGPU.matrix_gpu_copy (4 ms)
[ RUN      ] MathMatrixGPU.barebone_buffer_copy
[       OK ] MathMatrixGPU.barebone_buffer_copy (1 ms)
[----------] 3 tests from MathMatrixGPU (22 ms total)

[----------] Global test environment tear-down
[==========] 3 tests from 1 test case ran. (22 ms total)
[  PASSED  ] 2 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] MathMatrixGPU.matrix_gpu_copy

 1 FAILED TEST
```

File: stan/math/gpu/kernels/basic_matrix_kernels.cl [hunk 2]
```diff
@@ -0,0 +1,13 @@
+#define A(i,j)  A[j*rows+i]
```
Comment (syclik) [inline stan/math/gpu/kernels/basic_matrix_kernels.cl]: just curious: any reason this is macro'd instead of done directly? I know it's in the kernel code in the C-like language.

File: stan/math/gpu/matrix_gpu.hpp [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+
+/**
+ * This class represents a matrix on the GPU.
+ *
+ * The matrix data is stored in the oclBuffer_.
+ *
+ */
+class matrix_gpu {
+ private:
+  /** cl::Buffer provides functionality for working with OpenCL buffer.
```
Comment (syclik) [inline stan/math/gpu/matrix_gpu.hpp]: start this comment on the next line. so like:
```
/**
 * cl::Buffer...
```

File: stan/math/gpu/matrix_gpu.hpp [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+  int cols() const { return cols_; }
+
+  int size() const { return rows_ * cols_; }
+
+  const cl::Buffer& buffer() const { return oclBuffer_; }
+
+  matrix_gpu() : rows_(0), cols_(0) {}
+
+  matrix_gpu(const matrix_gpu& a) : rows_(a.rows()), cols_(a.cols()) {
+    // this check is needed because creating OpenCL
```
Comment (syclik) [inline stan/math/gpu/matrix_gpu.hpp]: I don't understand this comment. What does this mean? What creation are you referring to?

File: stan/math/gpu/matrix_gpu.hpp [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+        // creates a read&write object for "size" double values
+        // in the provided context
+        oclBuffer_
+            = cl::Buffer(ctx, CL_MEM_READ_WRITE, sizeof(double) * size());
+        /** Sets the arguments for the kernel. See copy_matrix_kernel in
+         * kernels/basic_matrix_gpu_kernels.hpp for the kernel code.
+         * The arguments are the source & destination matrices
+         * and the size in rows and columns.
+         */
+        kernel.setArg(0, a.buffer());
```
Comment (syclik) [inline stan/math/gpu/matrix_gpu.hpp]: Can we move the `cl::Kernel kernel = opencl_context.get_kernel("copy");` line to here right before the arguments are set? It'd be much easier to track what's going on.

If we can't, it would be really good to document why. I don't know if there's order dependency here.

File: stan/math/gpu/matrix_gpu.hpp [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+ * are stored on the GPU.
+ *
+ * @param src source matrix
+ * @param dst destination matrix
+ *
+ * @throw <code>std::invalid_argument</code> if the
+ * matrices do not have matching dimensions
+ *
+ */
+inline void copy(matrix_gpu& src, matrix_gpu& dst) {  // NOLINT
```
Comment (syclik) [inline stan/math/gpu/matrix_gpu.hpp]: why does this need a `// NOLINT` comment? If it's absolutely necessary, please include the specific exception to cpplint.

File: stan/math/gpu/opencl_context.hpp [hunk 7]
```diff
@@ -101,16 +103,32 @@ class opencl_context_base {
...
+  // Helper func for reading in kernels
+  inline const char* read_kernel(const char* file_name) {
+    std::ifstream kernel_buff(file_name);
+    static std::string kernel;
+    kernel_buff.seekg(0, std::ios::end);
+    kernel.reserve(kernel_buff.tellg());
+    kernel_buff.seekg(0, std::ios::beg);
+    kernel.assign((std::istreambuf_iterator<char>(kernel_buff)),
+                  std::istreambuf_iterator<char>());
+    return kernel.c_str();
```
Comment (syclik) [inline stan/math/gpu/opencl_context.hpp]: This return isn't safe. We ran into a problem with this years ago. See: https://stackoverflow.com/questions/22330250/how-to-return-a-stdstring-c-str

Mind checking if this is the `static` problem you saw?

Is it possible to operate using `std::string` where this is used?

**Comment by SteveBronder** [comment]
I'm 90% sure that's because of how I'm reading in the kernel. I'll fix that now

File: stan/math/gpu/matrix_gpu.hpp [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+ *
+ * @tparam T type of data in the Eigen matrix
+ * @param src source matrix on the GPU
+ * @param dst destination Eigen matrix
+ *
+ * @throw <code>std::invalid_argument</code> if the
+ * matrices do not have matching dimensions
+ *
+ */
+template <typename T, int R, int C>
```
Comment (syclik) [inline stan/math/gpu/matrix_gpu.hpp]: Will `T` ever be `stan::math::var`? 
- If not, we should take away the template parameter and make that `double`. 
- If so, we should test it just to make sure it works. I saw some manual allocation with `size_of(double)` -- just verifying that it does the right thing would be good.

File: stan/math/gpu/matrix_gpu.hpp [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+void copy(matrix_gpu& src, Eigen::Matrix<T, R, C>& dst) {
```
Comment (syclik) [inline stan/math/gpu/matrix_gpu.hpp]: These `copy` functions should all go into its own header in `stan/math/prim/mat/copy.hpp`.

**Comment by syclik** [comment]
Also: it needs copyright holder.

File: stan/math/gpu/kernels/basic_matrix_kernels.cl (hunk 2 above)
Comment (rok-cesnovar) [reply stan/math/gpu/kernels/basic_matrix_kernels.cl]: We use the same type of macros in the more complex kernels that will follow, to increase readabilitity. Its also used here for consistentcy purposes.

File: stan/math/gpu/matrix_gpu.hpp (hunk 4 above)
Comment (SteveBronder) [reply stan/math/gpu/matrix_gpu.hpp]: It's a two line comment
```
    // this check is needed because creating OpenCL
    // buffers of size 0 throws an OpenCL exception
```
OpenCL errors if you try to make a buffer of size 0

File: stan/math/gpu/matrix_gpu.hpp (hunk 8 above)
Comment (SteveBronder) [reply stan/math/gpu/matrix_gpu.hpp]: > Will T ever be stan::math::var?

No idt so. We can set that to double.

File: stan/math/gpu/matrix_gpu.hpp (hunk 6 above)
Comment (SteveBronder) [reply stan/math/gpu/matrix_gpu.hpp]: Removing the //NOLINT and running `make cpplint`

```
stan/math/gpu/matrix_gpu.hpp(252):  Add #include <algorithm> for copy  [build/include_what_you_use] [4]
```

Is there some way to turn the search for that word? I'm guessing we want that check elsewhere tho' so I can include a comment here on why we have a no lint

File: stan/math/gpu/matrix_gpu.hpp (hunk 8 above)
Comment (SteveBronder) [reply stan/math/gpu/matrix_gpu.hpp]: Actually. Today the answer is no. But in the future that could be a thing. 

@rok-cesnovar I think we would have to go through and template all of this, but how do you feel about passing vars onto the gpu?

Comment (rok-cesnovar) [reply stan/math/gpu/matrix_gpu.hpp]: Using C++ classes in kernel code is not yet supported (it was announced a month ago and it will be included in OpenCL 2.2), so we can't handle that for now. Even if we were to switch to 2.2, we would severely limit the number of supported devices if we did that.

Comment (SteveBronder) [reply stan/math/gpu/matrix_gpu.hpp]: Cool then let's remove the T and only accept accepting doubles (or we could limit T to just arithmetic types)

**Comment by SteveBronder** [comment]
@syclik I'm seeing a failure upstream I'm not sure how to interpret

http://d1m1s1b1.stat.columbia.edu:8080/job/Stan/job/downstream%20tests/421/execution/node/89/log/

```
g++  -Wall -I . -isystem lib/stan_math/lib/eigen_3.3.3 -isystem lib/stan_math/lib/boost_1.66.0 -isystem lib/stan_math/lib/sundials_3.1.0/include -std=c++1y -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION -Wno-unused-function -Wno-uninitialized -I src -isystem lib/stan_math/ -DFUSION_MAX_VECTOR_SIZE=12 -Wno-unused-local-typedefs -m64 -Wno-unused-but-set-variable -DNO_FPRINTF_OUTPUT -pipe   -O3 lib/stan_math/lib/gtest_1.7.0/src/gtest_main.cc test/unit/lang/parser/function_sig_errors.o -DGTEST_USE_OWN_TR1_TUPLE -DGTEST_HAS_PTHREAD=0 -isystem lib/stan_math/lib/gtest_1.7.0/include -isystem lib/stan_math/lib/gtest_1.7.0 -o test/unit/lang/parser/function_sig_errors.exe test/libgtest.a  -static-libgcc -static-libstdc++ -Lbin -lstanc 
test/unit/lang/parser/functions_standalone_instantiate.o:functions_standalone_instantiate_test.cpp:(.text$_ZN4stan4math17cvodes_integratorILi2EE9integrateIN19integrate_functions23integrand_ode_functor__EddEESt6vectorIS6_INS_11return_typeIT0_T1_ddddE4typeESaISB_EESaISD_EERKT_RKS6_IS8_SaIS8_EEdRKS6_IdSaIdEERKS6_IS9_SaIS9_EESQ_RKS6_IiSaIiEEPSoddl[_ZN4stan4math17cvodes_integratorILi2EE9integrateIN19integrate_functions23integrand_ode_functor__EddEESt6vectorIS6_INS_11return_typeIT0_T1_ddddE4typeESaISB_EESaISD_EERKT_RKS6_IS8_SaIS8_EEdRKS6_IdSaIdEERKS6_IS9_SaIS9_EESQ_RKS6_IiSaIiEEPSoddl]+0x3a1): undefined reference to `N_VMake_Serial'
```

`N_VMake_Serial' seems to be something from sundials?

**Comment by andrjohns** [comment]
I'm getting the same problem on an unrelated branch, it looks like a recent merge is giving the Stan repo some troubles. Running the ```[test/unit/lang/parser/functions_standalone_instantiate]``` unit test on the Stan develop branch gives the error.

File: stan/math/gpu/kernels/basic_matrix_kernels.cl (hunk 2 above)
Comment (syclik) [reply stan/math/gpu/kernels/basic_matrix_kernels.cl]: Thanks. That makes sense.

**Comment by SteveBronder** [comment]
Checks are passing now! @syclik do you want to try to run this again? If it doesn't work on your local we can chat tmrw to figure out why

**Comment by SteveBronder** [comment]
Also @davors is making some updates to the OpenCL wiki on our fork which looks helpful for getting setup!
https://github.com/bstatcomp/math/wiki/OpenCL-GPU-support

**Comment by syclik** [comment]
@SteveBronder, awesome! I just ran it and it now passes tests. Let me take another quick sweep.

File: stan/math/gpu/matrix_gpu.hpp [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+         * - the first argument is the kernel object
+         * - the second argument is the thread offset that is used for
+         *   calculating the global thread ID, NULL here, meaning the
+         *   offset is 0,0
+         * - the third argument specifies the amount of threads to create
+         *   in 2D (rows, columns)
+         * - the fourth argument specifies the size of the thread block,
+         *   NULL here, meaning the OpenCL driver determines the size
+         * - the last two arguments are for tracking events associated
+         *   with the kernel (enqueue,start, stop,...) for profiling.
```
Comment (syclik) [inline stan/math/gpu/matrix_gpu.hpp]: minor: consistent spacing. `enqueue, start, stop, ...` and remove the blank `*` line.

File: stan/math/gpu/matrix_gpu.hpp [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+    }
+  }
+  /**
+   * Constructor for the matrix_gpu that
+   * only allocates the buffer on the GPU.
+   *
+   * @param rows number of matrix rows
+   * @param cols number of matrix columns
+   *
+   * @throw <code>std::invalid_argument</code> if the
```
Comment (syclik) [inline stan/math/gpu/matrix_gpu.hpp]: This comment is not correct. Please update.

File: stan/math/gpu/matrix_gpu.hpp [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+   * matrices do not have matching dimensions
+   *
+   */
+  matrix_gpu(int rows, int cols) : rows_(rows), cols_(cols) {
+    if (size() > 0) {
```
Comment (syclik) [inline stan/math/gpu/matrix_gpu.hpp]: I think we should check the `rows` and `cols` here. This allows for both negative rows and negative cols.

I think the right thing to do is to use the `stan/math/prim/scal/err/check_positive_size.hpp` function to throw an `invalid_argument` exception here (the function automatically throws it).

File: stan/math/gpu/matrix_gpu.hpp [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+  }
+  /**
+   * Constructor for the matrix_gpu that
+   * creates a copy of the Eigen matrix on the GPU.
+   *
+   *
+   * @tparam T type of data in the Eigen matrix
+   * @param A the Eigen matrix
+   *
+   * @throw <code>std::invalid_argument</code> if the
```
Comment (syclik) [inline stan/math/gpu/matrix_gpu.hpp]: This comment isn't valid.

File: stan/math/gpu/matrix_gpu.hpp (hunk 12 above)
Comment (syclik) [reply stan/math/gpu/matrix_gpu.hpp]: I just read through the copy code. If you assume that the rows and cols are greater than or equal to 0, then I'd write it explicitly in the comment for `rows` and `cols`. It's ok if this code doesn't check it because it's checked outside, but it needs to be explicit that this is not going to check it.

File: stan/math/prim/mat/fun/opencl_copy.hpp [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+ * the destination matrix that is stored
+ * on the GPU.
+ *
+ * @tparam T type of data in the Eigen matrix
+ * @param src source Eigen matrix
+ * @param dst destination matrix on the GPU
+ *
+ * @throw <code>std::invalid_argument</code> if the
+ * matrices do not have matching dimensions
+ *
```
Comment (syclik) [inline stan/math/prim/mat/fun/opencl_copy.hpp]: remove blank comment line

File: stan/math/prim/mat/fun/opencl_copy.hpp [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+ */
+template <int R, int C>
+void copy(const Eigen::Matrix<double, R, C>& src, matrix_gpu& dst) {
```
Comment (syclik) [inline stan/math/prim/mat/fun/opencl_copy.hpp]: I'm really sorry I didn't catch this earlier. We want to switch the order of the arguments to be consistent with the rest of the Stan library. The first argument should be what gets filled in (destination in your language) and the second should be what is assigned (the source).

If you found this in another example, please let me know.

File: stan/math/prim/mat/fun/opencl_copy.hpp [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+ * @tparam T type of data in the Eigen matrix
+ * @param src source matrix on the GPU
+ * @param dst destination Eigen matrix
+ *
+ * @throw <code>std::invalid_argument</code> if the
+ * matrices do not have matching dimensions
+ *
+ */
+template <int R, int C>
+void copy(matrix_gpu& src, Eigen::Matrix<double, R, C>& dst) {
```
Comment (syclik) [inline stan/math/prim/mat/fun/opencl_copy.hpp]: Same comment here.

File: stan/math/prim/mat/fun/opencl_copy.hpp [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+ * are stored on the GPU.
+ *
+ * @param src source matrix
+ * @param dst destination matrix
+ *
+ * @throw <code>std::invalid_argument</code> if the
+ * matrices do not have matching dimensions
+ *
+ */
+inline void copy(matrix_gpu& src, matrix_gpu& dst) {
```
Comment (syclik) [inline stan/math/prim/mat/fun/opencl_copy.hpp]: Same comment as above.

File: stan/math/prim/mat/fun/opencl_copy.hpp [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+ * Copies the source matrix to the
+ * destination matrix. Both matrices
+ * are stored on the GPU.
+ *
+ * @param src source matrix
+ * @param dst destination matrix
+ *
+ * @throw <code>std::invalid_argument</code> if the
+ * matrices do not have matching dimensions
+ *
```
Comment (syclik) [inline stan/math/prim/mat/fun/opencl_copy.hpp]: remove blank comment line

File: test/unit/math/prim/mat/fun/opencl_copy_test.cpp [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+  EXPECT_EQ(4, d2_a(1, 0));
+  EXPECT_EQ(5, d2_a(1, 1));
+  EXPECT_EQ(6, d2_a(1, 2));
+  EXPECT_EQ(1, d2_b(0, 0));
+  EXPECT_EQ(2, d2_b(0, 1));
+  EXPECT_EQ(3, d2_b(0, 2));
+  EXPECT_EQ(4, d2_b(1, 0));
+  EXPECT_EQ(5, d2_b(1, 1));
+  EXPECT_EQ(6, d2_b(1, 2));
+}
```
Comment (syclik) [inline test/unit/math/prim/mat/fun/opencl_copy_test.cpp]: Mind adding a test for a `matrix_gpu` of size `(0, 0)`? We just want to make sure that edge case is handled correctly (starting now and for as long as we run the unit tests).

File: stan/math/gpu/kernels/basic_matrix_kernels.cl [hunk 20]
```diff
@@ -0,0 +1,14 @@
...
+#define B(i,j)  B[j*rows+i]
+ __kernel void copy(
+      __global double *A,
+      __global double *B,
+      unsigned int rows,
+      unsigned int cols) {
+  int i = get_global_id(0);
+  int j = get_global_id(1);
+  if ( i < rows && j < cols ) {
```
[the remark this answers is not in the document]
Comment (rok-cesnovar) [reply stan/math/gpu/kernels/basic_matrix_kernels.cl]: Done. I fixed the spacings, the missing spaces in (i,j) and the unnecessary spaces before __kernel.

File: stan/math/gpu/matrix_gpu.hpp [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+#define STAN_MATH_GPU_MATRIX_GPU_HPP
+#ifdef STAN_OPENCL
+#include <stan/math/gpu/opencl_context.hpp>
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/scal/err/check_size_match.hpp>
+#include <CL/cl.hpp>
+#include <iostream>
+#include <string>
+#include <vector>
+/*
```
[the remark this answers is not in the document]
Comment (rok-cesnovar) [reply stan/math/gpu/matrix_gpu.hpp]: Done.

File: stan/math/gpu/matrix_gpu.hpp [hunk 22]
```diff
@@ -0,0 +1,173 @@
...
+ */
+
+namespace stan {
+namespace math {
+
+/**
+ * This class represents a matrix on the GPU.
+ *
+ * The matrix data is stored in the oclBuffer_.
+ *
```
[the remark this answers is not in the document]
Comment (rok-cesnovar) [reply stan/math/gpu/matrix_gpu.hpp]: Done.

File: stan/math/gpu/matrix_gpu.hpp (hunk 10 above)
Comment (rok-cesnovar) [reply stan/math/gpu/matrix_gpu.hpp]: Done.

File: stan/math/gpu/matrix_gpu.hpp (hunk 11 above)
Comment (rok-cesnovar) [reply stan/math/gpu/matrix_gpu.hpp]: Done.

File: stan/math/gpu/matrix_gpu.hpp (hunk 12 above)
Comment (rok-cesnovar) [reply stan/math/gpu/matrix_gpu.hpp]: I updated the rows and cols comment lines. 
If I understand this correctly, this is the onyl correction here. Please correct me, If I missed something.

File: stan/math/gpu/matrix_gpu.hpp (hunk 13 above)
Comment (rok-cesnovar) [reply stan/math/gpu/matrix_gpu.hpp]: fixed to std::system_error

File: stan/math/prim/mat/fun/opencl_copy.hpp (hunk 14 above)
Comment (rok-cesnovar) [reply stan/math/prim/mat/fun/opencl_copy.hpp]: fixed

File: stan/math/prim/mat/fun/opencl_copy.hpp (hunk 18 above)
Comment (rok-cesnovar) [reply stan/math/prim/mat/fun/opencl_copy.hpp]: fixed

File: stan/math/prim/mat/fun/opencl_copy.hpp (hunk 17 above)
Comment (rok-cesnovar) [reply stan/math/prim/mat/fun/opencl_copy.hpp]: done

File: stan/math/prim/mat/fun/opencl_copy.hpp (hunk 16 above)
Comment (rok-cesnovar) [reply stan/math/prim/mat/fun/opencl_copy.hpp]: done

File: test/unit/math/prim/mat/fun/opencl_copy_test.cpp (hunk 19 above)
Comment (rok-cesnovar) [reply test/unit/math/prim/mat/fun/opencl_copy_test.cpp]: I added 2 tests: 1 in the copy_test to check that we do not throw on copies (Eigen->GPU, GPU->Eigen, GPU->GPU) and added a test to check creation of matrices of size 0.

**Comment by rok-cesnovar** [comment]
Thank you @syclik for the review comments. I made the requested changes.

Sorry for all the "fixed" and "done" spam. I just wanted to make sure I did not miss anything.

**Comment by syclik** [comment]
No problem! Thank you for being so quick! I'll take a look later tonight.

File: stan/math/prim/mat/fun/opencl_copy.hpp [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+namespace stan {
+namespace math {
+
+/**
+ * Copies the source Eigen matrix to
+ * the destination matrix that is stored
+ * on the GPU.
+ *
+ * @tparam T type of data in the Eigen matrix
+ * @param src source Eigen matrix
```
Comment (syclik) [inline stan/math/prim/mat/fun/opencl_copy.hpp]: Minor: these should match the order of the arguments. I'll try to fix it on your branch directly.

File: stan/math/prim/mat/fun/opencl_copy.hpp [hunk 24]
```diff
@@ -0,0 +1,132 @@
...
+ *
+ * @tparam T type of data in the Eigen matrix
+ * @param src source matrix on the GPU
+ * @param dst destination Eigen matrix
+ *
+ * @throw <code>std::invalid_argument</code> if the
+ * matrices do not have matching dimensions
+ */
+template <int R, int C>
+void copy(Eigen::Matrix<double, R, C>& dst, matrix_gpu& src) {
```
Comment (syclik) [inline stan/math/prim/mat/fun/opencl_copy.hpp]: const correctness on second arg. I'll try to fix this too.

File: stan/math/prim/mat/fun/opencl_copy.hpp [hunk 25]
```diff
@@ -0,0 +1,132 @@
...
+ * destination matrix. Both matrices
+ * are stored on the GPU.
+ *
+ * @param src source matrix
+ * @param dst destination matrix
+ *
+ * @throw <code>std::invalid_argument</code> if the
+ * matrices do not have matching dimensions
+ */
+inline void copy(matrix_gpu& dst, matrix_gpu& src) {
```
Comment (syclik) [inline stan/math/prim/mat/fun/opencl_copy.hpp]: const correctness

**Comment by syclik** [comment]
@rok-cesnovar, I was able to push directly to the branch! It looks good. We'll wait until tests pass and then merge.
